src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
 		NULL);
 }
 
+// !!!!!!! NOTE !!!!!!!!
+// The flags must match those in the ETW manifest exactly
+// !!!!!!! NOTE !!!!!!!!
+
+typedef enum {
+	K_ETW_TYPE_FLAGS_ARRAY = 0x8,
+} EtwTypeFlags;
+
+// This only contains the fixed-size data at the top of each struct in
+// the bulk type event.  These fields must still match exactly the initial
+// fields of the struct described in the manifest.
+typedef struct _EventStructBulkTypeFixedSizedData {
+	uint64_t type_id;
+    uint64_t module_id;
+    uint32_t type_name_id;
+    uint32_t flags;
+	uint8_t cor_element_type;
+} EventStructBulkTypeFixedSizedData;
+
+// Represents one instance of the Value struct inside a single BulkType event
+typedef struct _BultTypeValue {
+	EventStructBulkTypeFixedSizedData fixed_sized_data;
+	uint32_t c_type_parameters;
+	intptr_t rg_type_parameters[100]; // Should be variable length to save space
+	char *s_name;
+} BulkTypeValue;
+
+static
+uint32_t
+ep_rt_mono_get_byte_count_in_event(BulkTypeValue *bulk_type_value)
+{
+	return
+		sizeof(bulk_type_value->fixed_sized_data) +
+		sizeof(bulk_type_value->c_type_parameters) +
+		(sizeof(bulk_type_value->s_name) + 1) * sizeof(char) +  // Size of name, including null terminator
+		bulk_type_value->c_type_parameters * sizeof(uint64_t);	// Type parameters
+}
+
+// ETW has a limitation of 64K for TOTAL event Size, however there is overhead associated with
+// the event headers.   It is unclear exactly how much that is, but 1K should be sufficiently
+// far away to avoid problems without sacrificing the perf of bulk processing.
+static const uint32_t cb_max_etw_event = 63 * 1024;
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array.  (Intentionally overestimate the size of the non-array parts to keep it safe.)
+static const uint32_t k_max_bytes_type_values = (cb_max_etw_event - 0x30);
+
+// Estimate of how many type value elements we can put into the struct array, while
+// staying under the ETW event size limit. Note that this is impossible to calculate
+// perfectly, since each element of the struct array has variable size.
+//
+// In addition to the byte-size limit per event, Windows always forces on us a
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+static const uint32_t k_max_count_type_values = (128 - 2) / 4;
+
+BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+uint32_t m_n_bulk_type_value_count;
+uint32_t m_n_bulk_type_value_byte_count;
+uint8_t *m_p_bulk_type_event_buffer[65536];
```

Comment:
I think this should really be kept as local variables for each method detail events, bulk type emits all its types before sending the method details events, if you are worried about multiple reallocations of the buffer when adding bulk types, you could start with a local stack buffer (that covers most of the cases in its default size) and only if that is not enough move over to dynamic allocated buffer, but should not be needed frequently, unless the method uses a bunch of unique types.

Reply:
> also what do the prefixes mean?
I'm not actually sure, I was trying to keep it in parallel with CoreCLR. https://github.com/dotnet/runtime/blob/7d1836ac70233c62d71e2e8339eb9bf4a96457d9/src/coreclr/vm/eventtracepriv.h#L285-L289 but we can rename it if thats preferred.

Originally I just brought the variables in as globals to get things rolling. Thanks for reminding me of threading concerns.

Would we prefer to keep these as local variables for each method details events (would this look like passing around these as parameters?) or mimicking CoreCLR's BulkTypeEventLogger instance with a struct? I initially skipped attempting to implement CoreCLR's BulkTypeEventLogger instance, `TypeSystemLog`, `MethodLog` as I'm not very familiar with (C/C++) and just wanted to get the implementation going. I haven't grasped the broader intent with CoreCLR's structure, but is the current structure in mono roughly sufficient? Would it be better to make instances of each instead?

Reply:
Since we might end up using the bulk type event logger in more places going forward, I would do something similar as we done with other event pipe types, represent them as a struct and potentially add getter/setters init/fini and alloc/free (if you need heap allocation). Since this type is currently only used in Mono's EP implementation you should define this type within Mono specific EP sources, like ep-rt-mono.h

Reply:
Added structs for `BulkTypeEventLogger` that has `ep_rt_bulk_type_event_logger_init` and `ep_rt_bulk_type_event_logger_fini`, and `ep_rt_bulk_type_value_init` for the BulkTypeValues.